gobblin-runtime/src/main/java/gobblin/runtime/AbstractJobLauncher.java

```diff
@@ -116,97 +132,69 @@ public AbstractJobLauncher(Properties properties)
+    if (jobListener != null) {
+      jobListener.onJobCancellation(this.jobContext.getJobState());
     }
+  }
 
-    String jobId = jobProps.getProperty(ConfigurationKeys.JOB_ID_KEY);
-    // If no job ID is assigned (e.g., if the job is assigned through Azkaban), assign a new job ID here.
-    if (Strings.isNullOrEmpty(jobId)) {
-      jobId = JobLauncherUtils.newJobId(jobName);
-      jobProps.setProperty(ConfigurationKeys.JOB_ID_KEY, jobId);
-    }
-
-    State jobPropsState = new State();
-    // Add all job configuration properties of this job
-    jobPropsState.addAll(jobProps);
-
-
-    // Initialize the source for the job
-    JobState jobState;
-    Source<?, ?> source;
-    try {
-      JobState previousJobState = getPreviousJobState(jobName);
-      jobState = new JobState(jobPropsState, previousJobState.getTaskStatesAsWorkUnitStates(), jobName, jobId);
-      // Remember the number of consecutive failures of this job in the past
-      jobState.setProp(ConfigurationKeys.JOB_FAILURES_KEY,
-          previousJobState.getPropAsInt(ConfigurationKeys.JOB_FAILURES_KEY, 0));
-      source = new SourceDecorator(initSource(jobProps), jobId, LOG);
-    } catch (Throwable t) {
-      String errMsg = "Failed to initialize the source for job " + jobId;
-      LOG.error(errMsg + ": " + t, t);
-      unlockJob(jobName, jobLockOptional);
-      throw new JobException(errMsg, t);
+  @Override
+  @SuppressWarnings("unchecked")
+  public void launchJob(JobListener jobListener)
+      throws JobException {
+    if (!tryLockJob()) {
+      throw new JobException(
+          String.format("Previous instance of job %s is still running, skipping this scheduled run",
+              this.jobContext.getJobName()));
     }
 
-    JobMetrics jobMetrics = JobMetrics.get(jobState);
-    jobState.setProp(ConfigurationKeys.METRIC_CONTEXT_NAME_KEY, jobMetrics.getName());
-    jobState.setState(JobState.RunningState.PENDING);
+    String jobId = this.jobContext.getJobId();
+    JobState jobState = this.jobContext.getJobState();
 
     // Generate work units of the job from the source
-    Optional<List<WorkUnit>> workUnits = Optional.fromNullable(source.getWorkunits(jobState));
+    Optional<List<WorkUnit>> workUnits = Optional.fromNullable(this.jobContext.getSource().getWorkunits(jobState));
+    // The absence means there is something wrong getting the work units
     if (!workUnits.isPresent()) {
-      // The absence means there is something wrong getting the work units
-      source.shutdown(jobState);
-      unlockJob(jobName, jobLockOptional);
+      unlockJob();
       throw new JobException("Failed to get work units for job " + jobId);
     }
 
+    // No work unit to run
     if (workUnits.get().isEmpty()) {
-      // No real work to do
       LOG.warn("No work units have been created for job " + jobId);
-      source.shutdown(jobState);
-      unlockJob(jobName, jobLockOptional);
+      unlockJob();
```

Comment:
unlockJob() should be in a finally section. Having to call unlockJob in every exit branch is very error-prone.